lib/elaborate/elaborateReferences.ts

```diff
@@ -1,269 +1,361 @@
+      if (I.implementsIHasDeclarations(p) || I.implementsIHasStatements(p) || p instanceof O.OPackageInstantiation || p instanceof O.OContext) {
+        key = p;
+        break;
+      }
+    }
 
-      // package names are only referable in direct visibility
-      if (directlyVisible && (object instanceof OPackage || object instanceof OPackageBody || object instanceof OEntity || object instanceof OArchitecture)) {
-        this.evaluateDefinition(reference, object, false);
+    if (!this.scopeVisibilityMap.has(key)) {
+      this.fillVisibilityMap(key);
+    }
+    const list = this.scopeVisibilityMap.get(key);
+    if (list === undefined) {
+      throw new Error('no map found');
+    }
+    const searchText = reference.referenceToken.getLText();
+    if (fallback) {
+      const fallbackList = this.fallbackVisibilityMap.get(key);
+      if (fallbackList === undefined) {
+        throw new Error('no fallback map found');
       }
+      return (list.get(searchText) ?? []).concat(fallbackList.get(searchText) ?? []);
+    } else {
+      return list.get(searchText) ?? [];
+    }
+  }
+
 
+  castToRead(reference: O.OReference) {
+    if (reference instanceof O.OWrite === false) {
+      if (reference instanceof O.OSelectedName) {
+        Object.setPrototypeOf(reference, O.OSelectedNameRead.prototype);
+      } else {
+        Object.setPrototypeOf(reference, O.ORead.prototype);
+      }
     }
   }
 
-  elaborateSelectedNames(reference: OSelectedName | OSelectedNameWrite) {
-    const [libraryToken] = reference.prefixTokens;
-    let library: OLibrary | undefined;
-    for (const [obj] of scope(reference)) {
-      if (implementsIHasLibraries(obj)) {
-        for (const findLibrary of obj.libraries) {
-          if (findLibrary.lexerToken.getLText() == libraryToken.referenceToken.getLText()) {
-            library = findLibrary;
-          }
+  link(reference: O.OReference, obj: O.ObjectBase & (I.IHasReferenceLinks | I.IHasLabel)) {
+    reference.definitions.push(obj);
+    if (I.implementsIHasLabel(obj)) {
+      obj.labelLinks.push(reference);
+    } else {
+      obj.referenceLinks.push(reference);
+    }
+    this.castToRead(reference);
+  }
+
+  elaborateReference(reference: O.OReference) {
+    for (const obj of this.getList(reference)) { 
+      // if (obj instanceof O.OPackage) {
+      //   // packages can only be referenced with lib.pkg!
+      //   continue;
+      // }
+      // alias doesn't has aliasReferences but still referenceLinks
+      if (I.implementsIHasReference(obj) || obj instanceof O.OAlias || I.implementsIHasLabel(obj)) {
+        this.link(reference, obj);
+      }
+    }
+
+    // if nothing was found look in the fallback map
+    if (reference.definitions.length === 0) {
+      for (const obj of this.getList(reference, true)) {
+        // alias doesn't has aliasReferences but still referenceLinks
+        if (I.implementsIHasReference(obj) || obj instanceof O.OAlias) {
+          this.link(reference, obj);
```

Comment:
I am not sure, but how do we make sure, we do not get false positives?
Also if this finds more potentially correct results, could they be also correct in addition?
If this return correct results why not run it always?

Reply:
We can get false positives here but in a lot of cases we are just not able to infer the type correctly and my idea was to prefer false positives to false negatives. This is a similar behavior as before the refactor but we reduce the amount of false-positives by a lot by not using the fallback map when we already found one in the correct map.
And we do not want to run it always because the elabs above will never find a false-positive and,  therefore, if they found something, then this definitely is a definition.